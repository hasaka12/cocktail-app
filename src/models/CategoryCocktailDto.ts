export interface CategoryDto {
  strDrink: string;
  strDrinkThumb: string;
  idDrink: string;
}

export interface CategoryCocktailDto {
  drinks: CategoryDto[];
  key: string;
}
