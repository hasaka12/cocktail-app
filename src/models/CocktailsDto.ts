export interface CocktailsDto {
  idDrink: string;
  strAlcoholic: string;
  strCategory: string;
  strDrink: string;
  strDrinkThumb: string;
  strGlass: string;
  strInstructions: string;
}
