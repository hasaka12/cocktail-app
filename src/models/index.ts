export * from "./CocktailsDto";
export * from "./SearchCocktailDto";
