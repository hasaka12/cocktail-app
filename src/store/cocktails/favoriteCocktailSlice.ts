import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { CocktailsDto } from "../../models";

interface CounterState {
  data: CocktailsDto[];
}

const initialState: CounterState = {
  data: [],
};

const counterSlice = createSlice({
  name: "my-counter",
  initialState,
  reducers: {
    addFavoriteCocktails: (state, action: PayloadAction<CocktailsDto>) => {
      if (state.data.some((item) => item.idDrink === action.payload.idDrink)) {
        const index = state.data.findIndex(
          (item) => item.idDrink === action.payload.idDrink
        );
        state.data.splice(index, 1);
        return;
      }
      state.data.push(action.payload);
    },
  },
});

export const { addFavoriteCocktails } = counterSlice.actions;

export default counterSlice.reducer;
