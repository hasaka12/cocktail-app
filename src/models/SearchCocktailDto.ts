import { CocktailsDto } from "./CocktailsDto";

export interface SearchCocktailDto {
  drinks: CocktailsDto[];
}
