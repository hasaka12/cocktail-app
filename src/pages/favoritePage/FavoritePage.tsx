import { useDispatch, useSelector } from "react-redux";

import Typography from "@mui/material/Typography";

import CocktailCard from "../../components/CocktailCard/CocktailCard";
import { CocktailsDto } from "../../models";
import { AppDispatch, RootState } from "../../store/store";
import { addFavoriteCocktails } from "../../store/cocktails/favoriteCocktailSlice";

export default function FavoritePage() {
  const favoriteCocktail = useSelector(
    (state: RootState) => state.favoriteCocktail.data
  );
  const dispatch = useDispatch<AppDispatch>();

  const handleAddFavorite = (cocktail: CocktailsDto) => {
    dispatch(addFavoriteCocktails(cocktail));
  };

  return (
    <>
      <div className="flex justify-center mb-4">
        <Typography variant="h4" component="h4">
          Favorite Cocktails
        </Typography>
      </div>
      <div className="lg:grid lg:grid-cols-5 lg:gap-4">
        {favoriteCocktail && favoriteCocktail.length > 0
          ? favoriteCocktail.map((cocktail) => (
              <div
                key={cocktail.idDrink}
                className="flex lg:mx-2 my-2 justify-center items-center"
              >
                <CocktailCard
                  key={cocktail.idDrink}
                  title={cocktail.strDrink}
                  img={cocktail.strDrinkThumb}
                  desc={cocktail.strInstructions}
                  isAddedToFavorite
                  cocktail={cocktail}
                  handleAddFavorite={handleAddFavorite}
                />
              </div>
            ))
          : null}
      </div>
      {favoriteCocktail && favoriteCocktail.length === 0 ? (
        <Typography textAlign="center">No favorite cocktail</Typography>
      ) : null}
    </>
  );
}
