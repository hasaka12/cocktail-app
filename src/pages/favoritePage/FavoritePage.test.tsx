import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { Store, UnknownAction, configureStore } from "@reduxjs/toolkit";
import userEvent from "@testing-library/user-event";

import FavoritePage from "./FavoritePage";
import favoriteCocktailSlice, {
  addFavoriteCocktails,
} from "../../store/cocktails/favoriteCocktailSlice";
import { CocktailsDto } from "../../models";

const cocktail: CocktailsDto = {
  idDrink: "123",
  strAlcoholic: "Mojito",
  strCategory: "Alcoholic",
  strDrink: "Drink",
  strDrinkThumb: "thumb",
  strGlass: "Glass",
  strInstructions: "Instructions",
};

describe("FavoritePage", () => {
  let store: Store<unknown, UnknownAction, unknown>;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        favoriteCocktail: favoriteCocktailSlice,
      },
    });
  });

  test('renders "Favorite Cocktails" heading', () => {
    favoriteCocktailSlice({ data: [] }, addFavoriteCocktails(cocktail));
    const { asFragment, getByText } = render(
      <Provider store={store}>
        <FavoritePage />
      </Provider>
    );
    const headingElement = getByText(/Favorite Cocktails/i);
    expect(headingElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders no favorite cocktail message when there are no favorite cocktails", () => {
    render(
      <Provider store={store}>
        <FavoritePage />
      </Provider>
    );
    const noFavoriteMessage = screen.getByText(/No favorite cocktail/i);
    expect(noFavoriteMessage).toBeInTheDocument();
  });
});
