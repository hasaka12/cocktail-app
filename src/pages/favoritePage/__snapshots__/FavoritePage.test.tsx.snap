// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritePage renders "Favorite Cocktails" heading 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center mb-4"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Favorite Cocktails
    </h4>
  </div>
  <div
    class="lg:grid lg:grid-cols-5 lg:gap-4"
  />
  <p
    class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
  >
    No favorite cocktail
  </p>
</DocumentFragment>
`;
